provider.go

```diff
@@ -0,0 +1,237 @@
+
+	// credentials to authenticate in keycloak.
+	clientID     string
+	clientSecret string
+	realm        string
+
+	refreshThreshold time.Duration
+
+	tokenMu   *sync.RWMutex // guards token.
+	tokenCond *sync.Cond
+	token     *token // current active token.
+
+	stopOnce sync.Once
+	stopCh   chan struct{}
+}
+
+// New returns new Provider.
+func New(c Config) (*Provider, error) {
+	c.applyDefaults()
+
+	if err := c.validate(); err != nil {
+		return nil, errors.Wrap(err, "config validation failed")
+	}
+
+	mu := &sync.RWMutex{}
+
+	p := &Provider{
+		refreshThreshold: c.RefreshThreshold,
+		keycloak:         &keycloakWithMetrics{gocloak.NewClient(c.KeycloakHostPort)},
+		keycloakTimeout:  c.KeycloakTimeout,
+		clientID:         c.ClientID,
+		clientSecret:     c.ClientSecret,
+		realm:            c.Realm,
+		logger:           c.Logger,
+		tokenMu:          mu,
+		tokenCond:        sync.NewCond(mu),
+		stopCh:           make(chan struct{}),
+		stopOnce:         sync.Once{},
+	}
+
+	go p.updateLoop()
+
+	return p, nil
+}
+
+// setClient sets internal keycloak client.
+// must be used only for testing!
+func (p *Provider) setClient(c Keycloak) {
+	p.keycloak = c
+}
+
+func (p *Provider) updateLoop() {
+	onTokenUpdate := func(t *time.Timer, token *token) {
+		p.token = token
+		p.tokenCond.Broadcast()
+
+		if p.token != nil {
+			expiresIn := time.Duration(token.jwt.ExpiresIn) * time.Second
+			t.Reset(expiresIn - p.refreshThreshold)
+		} else {
+			t.Reset(100 * time.Millisecond)
+		}
+	}
+
+	t := time.NewTimer(0)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		case <-t.C:
+			func() {
+				ctx, cancel := context.WithTimeout(context.Background(), p.keycloakTimeout)
+				defer cancel()
+
+				p.tokenMu.Lock()
+				defer p.tokenMu.Unlock()
+
+				// if there is no active token then request new one.
+				if p.token == nil {
+					token, err := p.requestNewToken(ctx)
+					if err != nil {
+						p.logger.Error("failed to get keycloak token", zap.Error(err))
+						onTokenUpdate(t, nil)
+						return
+					}
+					p.logger.Debug("set new token", zap.String("token", token.jwt.AccessToken))
+					onTokenUpdate(t, token)
+					return
+				}
+
+				// refresh token
+				token, err := p.refreshToken(ctx, p.token)
+				if err != nil {
+					p.logger.Error("failed to refresh token", zap.Error(err))
+
+					onTokenUpdate(t, nil)
+					return
+				}
+
+				p.logger.Debug("refreshed token", zap.String("token", token.jwt.AccessToken))
+				onTokenUpdate(t, token)
+			}()
+		}
+	}
+}
+
+// Token returns an access token.
+func (p *Provider) Token() (*sarama.AccessToken, error) {
+	p.tokenMu.RLock()
+	token := p.token
+	p.tokenMu.RUnlock()
+
+	if token.activeNow() {
+		return token.toSaramaToken(), nil
+	}
+
+	start := time.Now()
+	p.tokenMu.Lock()
+	for !p.token.activeNow() {
+		p.tokenCond.Wait()
+
+		// we don't want to block forever here.
+		if time.Since(start) > p.keycloakTimeout {
+			return nil, context.DeadlineExceeded
```

Comment:
This may cause a panic in service which will use single config for different Kafka producers, it's better to `Unlock()` locker

Reply:
Good catch, thx!